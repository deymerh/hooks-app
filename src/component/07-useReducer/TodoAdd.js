import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, resetValues] = useForm({
    description: ''
  });
  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    };
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };
    handleAddTodo(newTodo);
    resetValues();
  };
  return (
    <>
      <h5>Agregar TODO</h5>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="Aprender..."
          className="form-control"
          autoComplete="off"
          onChange={handleInputChange}
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary btn-block mt-2"
        >
          Agregar
          </button>
      </form>
    </>
  );
};
