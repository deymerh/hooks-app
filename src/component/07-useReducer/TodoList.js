import React from 'react';
import { TodoListItem } from './TodoListItem';
import propTypes from 'prop-types';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map((todo, i) => (
          <TodoListItem
            key={todo.id}
            i={i}
            todo={todo}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        ))
      }
    </ul>
  );
};
TodoList.propTypes = {
  todos: propTypes.array.isRequired,
  handleDelete: propTypes.func.isRequired,
  handleToggle: propTypes.func.isRequired
};