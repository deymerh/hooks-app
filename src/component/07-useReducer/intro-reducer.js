const initialState = [{
  id: 1,
  task: 'Buy bread',
  done: false
}];

const taskReducer = (state = initialState, action) => {
  if (action?.type === 'AddTask') {
    return [...state, action.payload]
  }
  return state
};

let tasks = taskReducer();

//Nuevo objeto a agregar
const newTask = {
  id: 2,
  task: 'Run',
  done: false
}
//Accion a ejecutar
const AddTaskReducer = {
  type: 'AddTask',
  payload: newTask
}

tasks = taskReducer(tasks, AddTaskReducer)
console.log(tasks);