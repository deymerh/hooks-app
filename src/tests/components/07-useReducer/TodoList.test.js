import { shallow } from 'enzyme';
import { TodoList } from '../../../component/07-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en el <TodoList />', () => {

  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(<TodoList
    todos={demoTodos}
    handleDelete={handleDelete}
    handleToggle={handleToggle}
  />);

  test('Debe mostrar el componente correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe tener la cantidad de items envados como parametro', () => {
    expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
    expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
  });
});
