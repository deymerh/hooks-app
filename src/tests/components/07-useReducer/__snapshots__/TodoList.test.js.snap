// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <TodoList /> Debe mostrar el componente correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Node",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
